process/block/poolsCleaner/crossTxsPoolsCleaner.go

```diff
@@ -0,0 +1,285 @@
+		blockTransactionsPool:    dataPool.Transactions(),
+		rewardTransactionsPool:   dataPool.RewardTransactions(),
+		unsignedTransactionsPool: dataPool.UnsignedTransactions(),
+		rounder:                  rounder,
+		shardCoordinator:         shardCoordinator,
+	}
+
+	ctpc.mapCrossTxsRounds = make(map[string]*crossTxInfo)
+
+	ctpc.blockTransactionsPool.RegisterHandler(ctpc.receivedBlockTx)
+	ctpc.rewardTransactionsPool.RegisterHandler(ctpc.receivedRewardTx)
+	ctpc.unsignedTransactionsPool.RegisterHandler(ctpc.receivedUnsignedTx)
+
+	return &ctpc, nil
+}
+
+func (ctpc *crossTxsPoolsCleaner) receivedBlockTx(key []byte) {
+	if key == nil {
+		return
+	}
+
+	log.Trace("crossTxsPoolsCleaner.receivedBlockTx", "hash", key)
+	ctpc.processReceivedTx(key, blockTx)
+}
+
+func (ctpc *crossTxsPoolsCleaner) receivedRewardTx(key []byte) {
+	if key == nil {
+		return
+	}
+
+	log.Trace("crossTxsPoolsCleaner.receivedRewardTx", "hash", key)
+	ctpc.processReceivedTx(key, rewardTx)
+}
+
+func (ctpc *crossTxsPoolsCleaner) receivedUnsignedTx(key []byte) {
+	if key == nil {
+		return
+	}
+
+	log.Trace("crossTxsPoolsCleaner.receivedUnsignedTx", "hash", key)
+	ctpc.processReceivedTx(key, unsignedTx)
+}
+
+func (ctpc *crossTxsPoolsCleaner) processReceivedTx(key []byte, txType int8) {
+	isCrossTx := false
+	var crossTxInfo *crossTxInfo
+	for shardID := uint32(0); shardID < ctpc.shardCoordinator.NumberOfShards(); shardID++ {
+		isCrossTx, crossTxInfo = ctpc.isCrossTxFromShard(key, shardID, txType)
+		if isCrossTx {
+			break
+		}
+	}
+
+	if !isCrossTx {
+		isCrossTx, crossTxInfo = ctpc.isCrossTxFromShard(key, core.MetachainShardId, txType)
+		if !isCrossTx {
+			return
+		}
+	}
+
+	ctpc.mutMapCrossTxsRounds.Lock()
+	defer ctpc.mutMapCrossTxsRounds.Unlock()
+
+	if _, ok := ctpc.mapCrossTxsRounds[string(key)]; !ok {
+		crossTxInfo.round = ctpc.rounder.Index()
+		crossTxInfo.txType = txType
+		ctpc.mapCrossTxsRounds[string(key)] = crossTxInfo
+
+		log.Trace("transaction has been added",
+			"hash", key,
+			"round", crossTxInfo.round,
+			"sender shard", crossTxInfo.senderShardID,
+			"receiver shard", crossTxInfo.receiverShardID,
+			"type", getTxTypeName(crossTxInfo.txType))
+	}
+
+	ctpc.cleanCrossTxsPoolsIfNeeded()
+}
+
+func (ctpc *crossTxsPoolsCleaner) isCrossTxFromShard(txHash []byte, shardID uint32, txType int8) (bool, *crossTxInfo) {
+	transactionPool := ctpc.getTransactionPool(txType)
+	if transactionPool == nil {
+		return false, nil
+	}
+
+	selfShardID := ctpc.shardCoordinator.SelfId()
+	if shardID == selfShardID {
+		return false, nil
+	}
+
+	strCache := process.ShardCacherIdentifier(shardID, selfShardID)
+	txStore := transactionPool.ShardDataStore(strCache)
+	if txStore == nil {
+		return false, nil
+	}
+
+	_, ok := txStore.Peek(txHash)
+	if !ok {
+		return false, nil
+	}
+
+	return true, &crossTxInfo{txStore: txStore, senderShardID: shardID, receiverShardID: selfShardID}
+}
+
+func (ctpc *crossTxsPoolsCleaner) cleanCrossTxsPoolsIfNeeded() {
```

Comment:
wouldn't it be more efficient if you call this on a goroutine once in a while, instead of trying to remove every transaction after it is added ? For example do the checks for percentAllowed and MaxRoundsToKeepUnprocessedTransactions somewhere else and call the clean mechanism only when needed (maybe this function can be the handler for a `PoolAlmostFullNotifier` or something like this)

Reply:
Good point

Reply:
Refactored

Reply:
It is better now, but can you also implement a weighted semaphore instead of just a mutex ? 
Imagine this situation: you receive a bulk of 1000 transactions. all of them are valid. for each transaction, you will call the cleanup method which will clean older transactions. but you will clean the pool from the first transaction, the rest of 999 would probably call the cleanup method with no change. 
By using a semaphore, you guarantee that the cleanup method is used only once per bulk.

Reply:
Actually the cleanup... method is not called anymore from received... method (I have removed it from there), but only from this go routine and only once per minute. So no matter what it will clean all what is needed to be cleaned from one shot. One more thing is that this cleaner will remove txs based on their age and not on their count.

Reply:
You're right, sorry. It's ok